schedtool.8
.\" Text automatically generated by txt2man-1.4.5
.TH "SCHEDTOOL" 8 "1 November 2006" "" "System Manager's Manual"
.SH "NAME"
\fBschedtool \fP\- query and set CPU scheduling parameters

.SH "SYNOPSIS"
.nf 
.fam C
.TP 
\fBschedtool\fP
[\fB\-0\fP|\fB\-N\fP] [\fB\-1\fP|\fB\-F\fP] [\fB\-2\fP|\fB\-R\fP] [\fB\-3\fP|\fB\-B\fP] [\fB\-4\fP|\fB\-I\fP] [\fB\-5\fP|\fB\-D\fP]
[\fB\-M\fP \fIpolicy\fP]
[\fB\-a\fP \fIaffinity\fP] 
[\fB\-p\fP \fIprio\fP]
[\fB\-n\fP \fInice_level\fP]
[\fB\-e\fP \fIcommand [arg ...]\fP]
[\fB\-r\fP]
[\fB\-v\fP]
[\fB\-h\fP]
[LIST OF \fIPIDs\fP]
.fam T
.fi 

.SH "DESCRIPTION"
\fBschedtool\fP can set all CPU scheduling parameters Linux is capable of or display information for given processes.
.P
Long\-running, non\-interactive tasks may benefit from SCHED_BATCH as timeslices are longer, less system\-time is wasted by computing the next runnable process and the caches stay stable.
.P
Audio/video or other near\-realtime applications may run with less skipping if set to SCHED_RR.
Use the static priority\-switch \fB\-p\fP to designate inter\-process\-hierarchies.
.TP 
\fBschedtool\fP now supports setting the CPU\-affinity introduced in linux 2.5.8.
.SH "OPTIONS"
.TP 
.B 
\fB\-N\fP or \fB\-0\fP
set all \fIPIDs\fP to SCHED_NORMAL/OTHER
.TP 
.B 
\fB\-F\fP or \fB\-1\fP
to SCHED_FIFO   root\-credentials required
.TP 
.B 
\fB\-R\fP or \fB\-2\fP
to SCHED_RR     root\-credentials required
.TP 
.B 
\fB\-B\fP or \fB\-3\fP
to SCHED_BATCH  
.TP 
.B 
\fB\-I\fP or \fB\-4\fP
to SCHED_ISO	
.TP 
.B 
\fB\-D\fP or \fB\-5\fP
to SCHED_IDLEPRIO  
.TP
.B
\fB\-M\fP \fIpolicy\fP
for manual/raw mode; policy is the number of the scheduling policy (see above for 0-4). 
This option is mostly for kernel guys that want to test their new implementations.
.TP 
.B 
\fB\-p\fP \fIprio\fP
specify static priority \fBrequired\fP for SCHED_FIFO and SCHED_RR. Usually ranged from 1\-99.
.TP 
.B 
\fB\-a\fP \fIaffinity\fP
set the PID's \fIaffinity\fP to this bitmask (hexadecimal); alternatively, a list mode is supported.
.TP 
.B 
\fB\-n\fP \fInice_level\fP
set the PID's nice level; see \fBnice(2), nice(1)\fP.
.TP 
.B 
\fB\-e\fP \fIcommand [arg ...]\fP
execute \fIcommand\fP with given scheduling parameters (overwrites schedtool's process image). See EXAMPLES.
.TP 
.B 
\fB\-r\fP
display min and max priority for each policy.
.TP 
.B 
\fB\-v\fP
be verbose.
.TP 
.B 
\fB\-h\fP
help

.SH "EXAMPLES"
To query the $SHELL's policies:
.PP 
.nf 
.fam C
    #> schedtool $$

To query some \fIPIDs\fP, namely 1 2 and 3:
.PP 
.nf 
.fam C
    #> schedtool 1 2 3

.fam T
.fi 
To \fBexecute\fP mplayer in \fBSCHED_RR\fP with priority 20. The priority arg is needed
for both SCHED_RR and SCHED_FIFO.
.PP 
.nf 
.fam C
    #> schedtool \-R \-p 20 \-e mplayer -quiet some_file.avi

.fam T
.fi 
To set current shell to \fBSCHED_BATCH\fP, which all programs the shell starts will inherit:
.PP 
.nf 
.fam C
    #> schedtool \-3 $$

.fam T
.fi 
To set all processes with the name 'cpu_hog' to SCHED_BATCH:
.PP 
.nf 
.fam C
    #> schedtool \-3 `pidof cpu_hog`

.fam T
.fi 
To set a process' \fIaffinity\fP to only the first CPU (CPU0):
.PP 
.nf 
.fam C
    #> schedtool \-a 0x1 <PID>

.fam T
Using the list mode and affinty of CPU0 and CPU3:
.PP
.nf
.fam C
    #> schedtool -a 0,3 <PID>

.fam T
.fi 
A combination of an \fIaffinity\fP and a \fIpolicy\fP\-argument is \- of course \- always possible.
.PP 
.nf 
.fam C
   #> schedtool \-B \-a 0x1 <PID>

.fam T
.fi 

.SH "AFFINITY MASK"
The \fIaffinity\fP\-argument determines on which CPUs a process is allowed to run. It consists of a simple bitmask represented in hexadecimal.
CPU0 is denoted by the least\-significant bit, CPU1 by the second least\-significant and so on, thus giving:
.PP 
    \fB0x1\fP \-> only run on CPU0
.PP 
    \fB0x2\fP \-> only run on CPU1
.PP 
    \fB0x4\fP \-> only run on CPU2
.PP 
    \fB0x8\fP \-> only run on CPU3 ... and so on.
.PP 
Multi\-target CPUs may be specified using bitwise OR of the values:
.PP 
    \fB0x7\fP \-> run on CPUs 0, 1, 2 but \fBNOT\fP on 4
.PP 
    \fB0x3\fP \-> run only on CPUs 0 and 1
.PP 
The default is to run a process on all CPUs, giving a mask of
.PP 
    \fB0xf\fP for all 4 CPUs
.PP 
    \fB0xff\fP for all 8 CPUs

.SH "AFFINITY MASK - LIST MODE"
Alternatively a list mode is supported where you can specify the CPUs delimited by ",".
The following runs <PID> on CPU0 and CPU1 (equivalent to \fB0x3\fP):
.PP 
.nf 
.fam C
   #> schedtool \-a \fB0,1\fP <PID>

.SH "POLICY OVERVIEW"
\fBSCHED_NORMAL / SCHED_OTHER\fP
This is the default policy and for the average program with some interaction. Does preemption of other processes.

.PP
\fBSCHED_FIFO\fP
First-In, First Out Scheduler, used only for real-time constraints.
Processes in this class are usually not preempted by others, they need to free
themselves from the CPU via sched_yield() and as such you need special
designed applications. \fBUse with extreme care.\fP
\fBROOT-credentials required.\fP

.PP
\fBSCHED_RR\fP
Round-Robin Scheduler, also used for real-time constraints. CPU-time is assigned in an round-robin fashion
with a much smaller timeslice than with SCHED_NORMAL and processes in
this group are favoured over SCHED_NORMAL. Usable for audio/video applications near
peak rate of the system.
\fBROOT-credentials required.\fP

.PP
\fBSCHED_BATCH\fP [ since 2.6.16 in mainline ]
SCHED_BATCH was designed for non-interactive, CPU-bound applications.
It uses longer timeslices (to better exploit the cache), but can be
interrupted anytime by other processes in other classes to guaratee
interaction of the system. Processes in this class are selected last
but may result in a considerable speed-up (up to 300%). No interactive
boosting is done.

.PP
\fBSCHED_ISO\fP [ patch needed ]
SCHED_ISO was designed to give users a SCHED_RR-similar class.
To quote Con Kolivas: "This is a non-expiring scheduler policy designed to 
guarantee a timeslice within a reasonable latency while preventing starvation.
Good for gaming, video at the limits of hardware, video capture etc."

.PP
\fBSCHED_IDLEPRIO\fP [ patch needed ]
SCHED_IDLEPRIO is similar to SCHED_BATCH, but was explicitly designed
to consume only the time the CPU is idle. No interactive boosting is done.
If you used SCHED_BATCH in the -ck kernels this is what you want since
2.6.16

.SH "HINTS"
PID 0 means "current process", in our case, schedtool. May occur when using the \-e switch.
.PP 
Be careful with SCHED_FIFO! You may lock out other processes from the CPU, including your shell.
.PP 
For SCHED_BATCH you \fIcertainly\fP need the a recent 2.6 kernel.
.PP 
A short overview is given in SCHED_DESIGN and the \fBREADME\fP contains thourough discussion. The \fBINSTALL\fP file also lists all prerequisites and where you can get patches.
.PP 
Affinity 0x0 should never be used.
.SH "SEE ALSO"
\fBsched_setscheduler\fP(2), \fBsched_setaffinity\fP(2), \fBnice\fP(2), \fBnice\fP(1), \fBrenice\fP(3).

.SH "BUGS"
You need some knowledge about the kernel and scheduling. The author is a grumpy little elitist.

.SH "AUTHOR"
Freek
.P
Please contact me via freshmeat.net's "contact author"\-feature (http://freshmeat.net/projects/schedtool).


CHANGES
Tue Aug  6 14:42:36 CEST 2002
v0.3

-added SCHED_FIFO, SCHED_RR, static priorities
-updated documentation
-included some scheduler overview (sched-design.txt)
-misc typo-fixes
-added Makefile for fast compiling / installing
-fixed logical errors in CHECK_RANGE_PRIO, CHECK_RANGE_POLICY
	(must've been quite late ;-)


Sat Aug 10 16:26:19 CEST 2002
v0.5

-man-page is ready and also installed via make install
-README restructured
-Wondering about the fast version-progress?
The whole package is now roughly half of what I've been dreaming of.
Still missing are e.g. some numbers.


Fri Aug 30 16:52:27 CEST 2002
v0.6 (only for me)

-minor source (-h / -v switches), README and Makefile (target: zipman,
unzipman) fixes


Wed Sep  4 10:22:31 CEST 2002
v0.7

-huge README updates with some numbers, improvements
-return-value updates: return values represent the number of failed
sched_setscheduler()-calls with 0 being everything okay.


Sun Sep 22 00:18:09 CEST 2002
v0.8

-some code work, #define SCHED_*, replace 0 with SCHED_NORMAL and such minor
stuff
-clarified the meaning of "higher" in README
-asked for util-linux inclusion


Mon Nov 11 17:26:45 CET 2002
v0.8b

-minor cleanups


Tue Aug 26 19:53:41 CEST 2003
v0.9

-implement better error handling to return exactly (well I tried at my
best knowledge) what went wrong during the sched_setscheduler()-call
-set_process() no longer automatically calls print_process()
-minor README/man-page updates regarding kernel versions 2.5 and 2.6


Sun Aug 31 16:14:35 CEST 2003
v0.95

-add support for CPU-affinity
-simplify error-handling in code
-overhaul the man-page


Tue Dec 24 21:31:59 CET 2003 (Tue Jan  6 13:37:32 CET 2004)
v0.96

-it is bad practice to have the "prerequisites" containing technical
stuff before the "about / info"-section explaing what the hell this
even is. README changed accordingly.
-moved installing into file INSTALL
-Con made a suggestion for using schedtool as a wrapper to start processes
with adjusted scheduling-params. Here you go, Con (option -e for (e)xec)!
-added parenthesis for safety in print_process in the big 
if( ... || ... || ... ) --now--> if( (...) || (...) || (...) )
-fixed an argument-BUG ("prio >= prio_min" instead of "p >= ...") in 
#define CHECK_RANGE_PRIO which did not change behaviour, but was wrong.
-made my debug-printing accessible via -DDEBUG_ME
-merry christmas


Thu Jan 15 14:41:06 CET 2004
v0.97

-changed -e(xecute) semantics from fork() to in-place exec() much like RML's
tools do it. This means that the schedtool-process will be overwritten by 
the specified command.
-added padding for better readable output; may change to dynamic spacing in 
future versions


Wed Jan 21 15:44:59 CET 2004
v0.98

-there, people, you have it: your changed cmdline parsing, no need for quoting
multi-argument-commands, just plain 
#> schedtool -3 -e cpu-hog -x1 -l2
THIS MEANS ALSO THAT YOU NO LONGER CAN MIX ARGS AND OPTIONS LIKE:
#> schedtool 1 2 3 4 -3 5 6
-restructured the help-output to be more "comprehendable" ... as an admin I get
serious pain reading it now
-and, hooray, schedtool is now silent "as that is standard unix behaviour"
unless you specify -v, which will print for each setting PID and so on, which
is indeed what I prefer to have.

Now that you have it all like you wanted, 
	send me money or even better, hot chicks.


Wed Jan 28 13:51:05 CET 2004
v0.99

-build schedtool with affinity is now the default
-changed Makefile a bit; new target no_affinity
-included support for SCHED_ISO (patch by Con Kolivas)
-update docs to reflect changes


Sun Apr  4 10:38:17 CEST 2004
v0.99.1

-support nice operations; schedtool is now capable of setting ALL scheduling
attributes!
-minor change in engine (move execute into the loop to eliminate duplicate 
code; the code is slowly getting a mess with all these checks here and there)
-documentation /man-page update; man-page now lists SCHED_*
-Makefile: new target install-doc, installs some docs into $PREFIX/share/doc/schedtool


Tue Apr 20 11:45:19 CEST 2004
v1.0

-released 1.0; more documentation updates and clarifications


Wed Jun 30 20:27:29 CEST 2004
v1.1.0

-added "NO WARRANTY" to help output
-add raw mode for the kernel hackers; allows direct input of the numerical 
policy value and changed docs accordingly.


Mon Jul  5 14:12:35 CEST 2004
v1.1.1

-added a list mode for affinity, additional to the bitmask mode
-identified 2 FIXMEs


Sun Aug  1 14:00:58 CEST 2004

-fixed one fixme: check affinity input for isdigit()
-changed the CHECK_RANGE_PRIO macro a bit for code cleanup
-moved affinity setting to the end of the loop, because may fail with ENOSYS


Sun Sep 19 19:50:09 CEST 2004
v1.1.2

-added missing padding for NICE in output
-changed custom "policy not implemented" error to "value out of range / policy
not implemented" to indicate wrong values when setting affinity or such


Tue Sep 28 20:59:42 CEST 2004
v1.2

-finally try to not break affinity-compiled binaries on non-affinity kernels
(2nd FIXME fixed)
-use getpid() instead of PID 0 (== current process) when in execute mode


Sun Oct  3 18:14:46 CEST 2004
v1.2.1

-REALLY DO NOT break affinity stuff; sched_getaffinity() suddenly returns 4 on
success instead of 0 as declared in the man-page on at least kernels 2.6.8+
-pad nice-levels correctly; they can be negative, as such we need %3d in 
printf()


Tue Oct 12 13:04:53 CEST 2004
v1.2.2

-adjust copyright
-correct typo in man-page
-correctly figure out affinity; I was missing an #include <linux/unistd.h>


Sun Oct 17 00:41:43 CEST 2004
v1.2.3

-get the affinity-ifdef-stuff right. Oh my.
-really up version to 1.2.3 (I'm getting old)
-remove superfluous break in schedtool.c::engine(), execute part
-correct another typo in man-page


Sat Nov 27 00:27:23 CET 2004
v1.2.4

-add a "probe" mode (-r), similar to sched-utils; lists prio_min/max
-add helper-functions get_prio_min_max() and print_prio_min_max()


Sat Feb  5 19:59:44 CET 2005
v1.2.5


schedtool.c
/*
 schedtool
 Copyright (C) 2002-2006 by Freek
 Please contact me via freshmeat.net
 Release under GPL, version 2
 Use at your own risk.
 Inspired by setbatch (C) 2002 Ingo Molnar

 01/2006:
 included support for SCHED_IDLEPRIO for ck kernels

 01/2004:
 included SCHED_ISO patch by Con Kolivas

 11/2004:
 add probing for some features (priority)

 09/2008:
 change affinity calls to new cpu_set_t API


 Born in the need of querying and setting SCHED_* policies.
 All output, even errors, go to STDOUT to ease piping.


 Content:

 main code
 cmd-line parsing
 the engine
 set_/print_process
 usage

 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sched.h>
#include <unistd.h>
#include <stdint.h>

#include "error.h"
#include "util.h"


/* various operation modes: print/set/affinity/fork */
#define MODE_NOTHING	0x0
#define MODE_PRINT	0x1
#define MODE_SETPOLICY	0x2
#define MODE_AFFINITY	0x4
#define MODE_EXEC	0x8
#define MODE_NICE       0x10
#define VERSION "1.3.0"

/*
 constants are from the O(1)-sched kernel's include/sched.h
 I don't want to include kernel-headers.
 Included those defines for improved readability.
 */
#undef SCHED_NORMAL
#undef SCHED_FIFO
#undef SCHED_RR
#undef SCHED_BATCH
#define SCHED_NORMAL	0
#define SCHED_FIFO	1
#define SCHED_RR	2
#define SCHED_BATCH	3
#define SCHED_ISO	4
#define SCHED_IDLEPRIO	5

/* for loops */
#define SCHED_MIN SCHED_NORMAL
#define SCHED_MAX SCHED_IDLEPRIO

#define CHECK_RANGE_POLICY(p) (p <= SCHED_MAX && p >= SCHED_MIN)
#define CHECK_RANGE_NICE(n) (n <= 20 && n >= -20)

/*
 4 bits of CPU_SETSIZE will naturally reduce to 1 char, so
 we'd only need [CPU_SETSIZE / 4 + 1]
 to be sure leave a bit of room and only do CPU_SETSIZE / 2
 */
#define CPUSET_HEXSTRING(name) char name[CPU_SETSIZE / 2]

char *TAB[] = {
	"N: SCHED_NORMAL",
	"F: SCHED_FIFO",
	"R: SCHED_RR",
	"B: SCHED_BATCH",
	"I: SCHED_ISO",
	"D: SCHED_IDLEPRIO",
	0
};

/* call it engine_s in lack of a better name */
struct engine_s {

	int mode;
	int policy;
	int prio;
        int nice;
	cpu_set_t aff_mask;

	/* # of args when going in PID-mode */
	int n;
	char **args;
};


int engine(struct engine_s *e);
int set_process(pid_t pid, int policy, int prio);
static inline int val_to_char(int v);
static char * cpuset_to_str(cpu_set_t *mask, char *str);
static inline int char_to_val(int c);
static int str_to_cpuset(cpu_set_t *mask, const char* str);
int parse_affinity(cpu_set_t *, char *arg);
int set_affinity(pid_t pid, cpu_set_t *mask);
int set_niceness(pid_t pid, int nice);
void probe_sched_features();
void get_prio_min_max(int policy, int *min, int *max);
void print_prio_min_max(int policy);
void print_process(pid_t pid);
void usage(void);


extern char *optarg;
extern int optind, opterr, optopt;
extern int errno;

int main(int ac, char **dc)
{
	/*
	 policy: -1, to indicate it was not set;
	 nice: 10, as nice/renice have as default;
	 prio: 0 per default
	 mode: MODE_NOTHING, no options set
	 */
	int policy=-1, nice=10, prio=0, mode=MODE_NOTHING;

	/*
	 aff_mask: zero it out
	 */
	cpu_set_t aff_mask;
        CPU_ZERO(&aff_mask);

        /* for getopt() */
	int c;

	if (ac < 2) {
		usage();
		return(0);
	}

	while((c=getopt(ac, dc, "+NFRBID012345M:a:p:n:ervh")) != -1) {

		switch(c) {
		case '0':
		case 'N':
			policy=SCHED_NORMAL;
                        mode |= MODE_SETPOLICY;
			break;
		case '1':
		case 'F':
			policy=SCHED_FIFO;
                        mode |= MODE_SETPOLICY;
			break;
		case '2':
		case 'R':
			policy=SCHED_RR;
                        mode |= MODE_SETPOLICY;
			break;
		case '3':
		case 'B':
			policy=SCHED_BATCH;
                        mode |= MODE_SETPOLICY;
			break;
		case '4':
		case 'I':
			policy=SCHED_ISO;
                        mode |= MODE_SETPOLICY;
			break;
		case '5':
		case 'D':
			policy=SCHED_IDLEPRIO;
                        mode |= MODE_SETPOLICY;
			break;
		case 'M':
			/* manual setting */
			policy=atoi(optarg);
			mode |= MODE_SETPOLICY;
			break;
		case 'a':
			mode |= MODE_AFFINITY;
			parse_affinity(&aff_mask, optarg);
                        break;
		case 'n':
                        mode |= MODE_NICE;
			nice=atoi(optarg);
                        break;
		case 'e':
			mode |= MODE_EXEC;
			break;
		case 'p':
			prio=atoi(optarg);
			break;
		case 'r':
                        probe_sched_features();
			break;
		case 'v':
                        /* the user wants feedback for each process */
			mode |= MODE_PRINT;
                        break;
		case 'V':
		case 'h':
			usage();
                        return(0);
		default:
			//printf("ERROR: unknown switch\n");
			usage();
			return(1);
		}
	}

	/*
         DAMN FUCKING
	 parameter checking
         ARGH!
	 */
	/* for _BATCH and _NORMAL, prio is ignored and must be 0*/
	if((policy==SCHED_NORMAL || policy==SCHED_BATCH) && prio) {
		decode_error("%s call may fail as static PRIO must be 0 or omitted",
			     TAB[policy]
			    );

	/* _FIFO and _RR MUST have prio set */
	} else if((policy==SCHED_FIFO || policy==SCHED_RR || policy==SCHED_ISO)) {

#define CHECK_RANGE_PRIO(p, p_low, p_high) (p <= (p_high) && p >= (p_low))
		/* FIFO and RR - check min/max priority */
		int prio_min, prio_max;

		get_prio_min_max(policy, &prio_min, &prio_max);
		//int prio_max=sched_get_priority_max(policy);
		//int prio_min=sched_get_priority_min(policy);

		if(! CHECK_RANGE_PRIO(prio, prio_min, prio_max)) {
			// this could be problematic on very special custom kernels
			if(prio == 0) {
				decode_error("missing priority; specify static priority via -p");
			} else {
				decode_error("PRIO %d is out of range %d-%d for %s",
					     prio,
					     prio_min,
					     prio_max,
					     TAB[policy]
					    );
			}
			/* treat as all calls have failed */
			return(ac-optind);
		}
#undef CHECK_RANGE_PRIO
	}

	/* no mode -> do querying */
	if(! mode) {
		mode |= MODE_PRINT;
	}

	if( mode_set(mode, MODE_EXEC) && ! ( mode_set(mode, MODE_SETPOLICY)
					     || mode_set(mode, MODE_AFFINITY)
                                             || mode_set(mode, MODE_NICE) )

	  ) {
		/* we have nothing to do */
		decode_error("Option -e needs scheduling-parameters, not given - exiting");
                return(-1);
	}

	if(! CHECK_RANGE_NICE(nice)) {
		decode_error("NICE %d is out of range -20 to 20", nice);
                return(-1);
	}
#undef CHECK_RANGE_NICE

	/* and: ignition */
	{
                struct engine_s stuff;
		/* now fill struct */
		stuff.mode=mode;
		stuff.policy=policy;
		stuff.prio=prio;
		stuff.aff_mask=aff_mask;
		stuff.nice=nice;

                /* we have this much real args/PIDs to process */
		stuff.n=ac-optind;

                /* this is the first real arg */
		stuff.args=dc+optind;

		/* now go on and do what we were told */
		return(engine(&stuff));
	}
}


int engine(struct engine_s *e)
{
	int ret=0;
	int i;

#ifdef DEBUG
	do {
		CPUSET_HEXSTRING(tmpaff);
		printf("Dumping mode: 0x%x\n", e->mode);
		printf("Dumping affinity: 0x%s\n", cpuset_to_str(&(e->aff_mask), tmpaff));
		printf("We have %d args to do\n", e->n);
		for(i=0;i < e->n; i++) {
			printf("Dump arg %d: %s\n", i, e->args[i]);
		}
	} while(0);
#endif

	/*
	 handle normal query/set operation:
	 set/query all given PIDs
	 */
	for(i=0; i < e->n; i++) {

		int pid, tmpret=0;
		cpu_set_t affi;

		CPU_ZERO(&affi);
                CPU_SET(0, &affi);

                /* if in MODE_EXEC skip check for PIDs */
		if(mode_set(e->mode, MODE_EXEC)) {
			pid=getpid();
			goto exec_mode_special;
		}

		if(! (isdigit( *(e->args[i])) ) ) {
			decode_error("Ignoring arg %s: is not a PID", e->args[i]);
			continue;
		}

		pid=atoi(e->args[i]);

	exec_mode_special:
		if(mode_set(e->mode, MODE_SETPOLICY)) {
			/*
			 accumulate possible errors
			 the return value of main will indicate
			 how much set-calls went wrong
                         set_process returns -1 upon failure
			 */
			tmpret=set_process(pid,e->policy,e->prio);
			ret += tmpret;

                        /* don't proceed as something went wrong already */
			if(tmpret) {
				continue;
			}

		}

		if(mode_set(e->mode, MODE_NICE)) {
			tmpret=set_niceness(pid, e->nice);
                        ret += tmpret;

			if(tmpret) {
				continue;
			}

		}

		if(mode_set(e->mode, MODE_AFFINITY)) {
			tmpret=set_affinity(pid, &(e->aff_mask));
			ret += tmpret;

			if(tmpret) {
				continue;
			}

		}

		/* and print process info when set, too */
		if(mode_set(e->mode, MODE_PRINT)) {
			print_process(pid);
		}


		/* EXECUTE: at the end */
		if(mode_set(e->mode, MODE_EXEC)) {

			char **new_argv=e->args;

			ret=execvp(*new_argv, new_argv);

			/* only reached on error */
			decode_error("schedtool: Could not exec %s", *new_argv);
			return(ret);
		}
	}
	/*
	 indicate how many errors we got; as ret is accumulated negative,
	 convert to positive
	 */
	return(abs(ret));
}


int set_process(pid_t pid, int policy, int prio)
{
	struct sched_param p;
	int ret;

	char *msg1="could not set PID %d to %s";
	char *msg2="could not set PID %d to raw policy #%d";

	p.sched_priority=prio;

	/* anything other than 0 indicates error */
	if((ret=sched_setscheduler(pid, policy, &p))) {

                /* la la pointer mismatch .. lala */
		decode_error((CHECK_RANGE_POLICY(policy) ? msg1 : msg2),
			     pid,
			     (CHECK_RANGE_POLICY(policy) ? TAB[policy] : policy)
			    );
		return(ret);
	}
	return(0);
}


/*
 the following functions have been taken from taskset of util-linux
 (C) 2004 by Robert Love
 */
static inline int val_to_char(int v)
{
	if (v >= 0 && v < 10)
		return '0' + v;
	else if (v >= 10 && v < 16)
		return ('a' - 10) + v;
	else 
		return -1;
}


/*
 str seems to need to hold [CPU_SETSIZE * 7] chars, as in used in taskset
 however, 4 bits of CPU_SETSIZE will naturally reduce to 1 char
 (bits "1111" -> char "f") so CPU_SETSIZE / 4 + 1 should be sufficient

 will pad with zeroes to the start, so print the string starting at the
 returned char *
 */
static char * cpuset_to_str(cpu_set_t *mask, char *str)
{
	int base;
	char *ptr = str;
	char *ret = 0;

	for (base = CPU_SETSIZE - 4; base >= 0; base -= 4) {
		char val = 0;
		if (CPU_ISSET(base, mask))
			val |= 1;
		if (CPU_ISSET(base + 1, mask))
			val |= 2;
		if (CPU_ISSET(base + 2, mask))
			val |= 4;
		if (CPU_ISSET(base + 3, mask))
			val |= 8;
		if (!ret && val)
			ret = ptr;
		*ptr++ = val_to_char(val);
	}
	*ptr = 0;
	return ret ? ret : ptr - 1;
}


static inline int char_to_val(int c)
{
	int cl;

	cl = tolower(c);
	if (c >= '0' && c <= '9')
		return c - '0';
	else if (cl >= 'a' && cl <= 'f')
		return cl + (10 - 'a');
	else
		return -1;
}


static int str_to_cpuset(cpu_set_t *mask, const char* str)
{
	int len = strlen(str);
	const char *ptr = str + len - 1;
	int base = 0;

	/* skip 0x, it's all hex anyway */
	if(len > 1 && str[0] == '0' && str[1] == 'x') {
		str += 2;
	}

	CPU_ZERO(mask);
	while (ptr >= str) {
		char val = char_to_val(*ptr);
		if (val == (char) -1)
			return -1;
		if (val & 1)
			CPU_SET(base, mask);
		if (val & 2)
			CPU_SET(base + 1, mask);
		if (val & 4)
			CPU_SET(base + 2, mask);
		if (val & 8)
			CPU_SET(base + 3, mask);
		len--;
		ptr--;
		base += 4;
	}

	return 0;
}
/* end of functions taken */


/* mhm - we need something clever for all that CPU_SET() and CPU_ISSET() stuff */
int parse_affinity(cpu_set_t *mask, char *arg)
{
	cpu_set_t tmp_aff;
	char *tmp_arg;
	size_t valid_len;

        CPU_ZERO(&tmp_aff);

	if(*arg == '0' && *(arg+1) == 'x') {
		/* we're in standard hex mode */
		str_to_cpuset(&tmp_aff, arg);

	} else if( (valid_len=strspn(arg, "0123456789,.")) ) {
		/* new list mode: schedtool -a 0,2 -> run on CPU0 and CPU2 */

		/* split on ',' and '.', because '.' is near ',' :) */
		while((tmp_arg=strsep(&arg, ",."))) {
                        int tmp_cpu;

			if(isdigit((int)*tmp_arg)) {
				tmp_cpu=atoi(tmp_arg);
                                CPU_SET(tmp_cpu, &tmp_aff);
#ifdef DEBUG
				printf("tmp_arg: %s -> tmp_cpu: %d\n", tmp_arg, tmp_cpu);
#endif
			}
		}

	} else {
		decode_error("affinity %s is not parseable", arg);
		exit(1);
	}

        *mask=tmp_aff;
	return 0;
}


int set_affinity(pid_t pid, cpu_set_t *mask)
{
	int ret;
	CPUSET_HEXSTRING(aff_hex);

	if((ret=sched_setaffinity(pid, sizeof(cpu_set_t), mask)) == -1) {
		decode_error("could not set PID %d to affinity 0x%s",
			     pid,
			     cpuset_to_str(mask, aff_hex)
			    );
		return(ret);
	}
        return(0);
}


int set_niceness(pid_t pid, int nice)
{
	int ret;

	if((ret=setpriority(PRIO_PROCESS, pid, nice))) {
		decode_error("could not set PID %d to nice %d",
			     pid,
			     nice
			    );
                return(ret);
	}
        return(0);
}


/*
 probe some features; just basic right now
 */
void probe_sched_features()
{
	int i;

	for(i=SCHED_MIN; i <= SCHED_MAX; i++) {
		print_prio_min_max(i);
 	}
}


/*
 get the min/max static priorites of a given policy. Max only work
 for SCHED_FIFO / SCHED_RR
 */
void get_prio_min_max(int policy, int *min, int *max)
{
	*min=sched_get_priority_min(policy);
        *max=sched_get_priority_max(policy);
}


/* print the min and max priority of a given policy, like chrt does */
void print_prio_min_max(int policy)
{
	int min, max;

	get_prio_min_max(policy, &min, &max);

	switch(min|max) {

	case -1:
		printf("%-17s: policy not implemented\n", TAB[policy]);
                break;
	default:
		printf("%-17s: prio_min %d, prio_max %d\n", TAB[policy], min, max);
                break;
	}
}


/*
 Be more careful with at least the affinity call; someone may use an
 affinity-compiled version on a non-affinity kernel.
 This is getting more and more fu-gly.
 */
void print_process(pid_t pid)
{
	int policy, nice;
	struct sched_param p;
	cpu_set_t aff_mask;
	CPUSET_HEXSTRING(aff_mask_hex);


	CPU_ZERO(&aff_mask);

	/* strict error checking not needed - it works or not. */
        errno=0;
	if( ((policy=sched_getscheduler(pid)) < 0)
	    || (sched_getparam(pid, &p) < 0)
	    /* getpriority may successfully return negative values, so errno needs to be checked */
	    || ((nice=getpriority(PRIO_PROCESS, pid)) && errno)
	  ) {
		decode_error("could not get scheduling-information for PID %d", pid);

	} else {

		/* do custom output for unknown policy */
		if(! CHECK_RANGE_POLICY(policy)) {
			printf("PID %5d: PRIO %3d, POLICY %-5d <UNKNOWN>, NICE %3d",
			       pid,
			       p.sched_priority,
			       policy,
			       nice
			      );
		} else {

			printf("PID %5d: PRIO %3d, POLICY %-17s, NICE %3d",
			       pid,
			       p.sched_priority,
			       TAB[policy],
			       nice
			      );
		}

		/*
		 sched_getaffinity() seems to also return (int)4 on 2.6.8+ on x86 when successful.
		 this goes against the documentation
                 */
		if(sched_getaffinity(pid, sizeof(aff_mask), &aff_mask) == -1) {
			/*
			 error or -ENOSYS
                         simply ignore and reset errno!
			 */
                        errno=0;
		} else {
			printf(", AFFINITY 0x%s", cpuset_to_str(&aff_mask, aff_mask_hex));
		}
	printf("\n");
	}
}


void usage(void)
{
	printf(
               "get/set scheduling policies - v" VERSION ", GPL'd, NO WARRANTY\n" \
               "USAGE: schedtool PIDS                    - query PIDS\n" \
               "       schedtool [OPTIONS] PIDS          - set PIDS\n" \
               "       schedtool [OPTIONS] -e COMMAND    - exec COMMAND\n" \
               "\n" \
               "set scheduling policies:\n" \
               "    -N                    for SCHED_NORMAL\n" \
               "    -F -p PRIO            for SCHED_FIFO       only as root\n" \
               "    -R -p PRIO            for SCHED_RR         only as root\n" \
               "    -B                    for SCHED_BATCH\n" \
               "    -I -p PRIO            for SCHED_ISO\n" \
               "    -D                    for SCHED_IDLEPRIO\n" \
               "\n" \
               "    -M POLICY             for manual mode; raw number for POLICY\n" \
               "    -p STATIC_PRIORITY    usually 1-99; only for FIFO or RR\n" \
               "                          higher numbers means higher priority\n" \
               "    -n NICE_LEVEL         set niceness to NICE_LEVEL\n" \
               "    -a AFFINITY_MASK      set CPU-affinity to bitmask or list\n\n" \
               "    -e COMMAND [ARGS]     start COMMAND with specified policy/priority\n" \
               "    -r                    display priority min/max for each policy\n" \
               "    -v                    be verbose\n" \
	       "\n" \
	      );
/*	printf("Parent ");
	print_process(getppid()); */

}
